File: alg_ds_1/ordered_list/ordered_list-2.go
package main

import (
	"constraints"
	"errors"
	"fmt"
	"os"
)

func (l *OrderedList[T]) RemoveDuplicates() {
	if l.head == nil {
		return
	}

	current := l.head
	for current != nil && current.next != nil {
		if l.Compare(current.value, current.next.value) == 0 {
			nodeToRemove := current.next
			current.next = nodeToRemove.next
			if nodeToRemove.next != nil {
				nodeToRemove.next.prev = current
			} else {
				l.tail = current
			}
		} else {
			current = current.next
		}
	}
}

func (l *OrderedList[T]) Merge(other *OrderedList[T]) error {
	if l._ascending != other._ascending {
		return errors.New("списки должны иметь одинаковый порядок сортировки")
	}

	if other.head == nil {
		return nil
	}

	if l.head == nil {
		l.head = other.head
		l.tail = other.tail
		return nil
	}

	merged := &OrderedList[T]{}
	merged.Clear(l._ascending)

	current1 := l.head
	current2 := other.head
	for current1 != nil && current2 != nil {
		if (l._ascending && l.Compare(current1.value, current2.value) <= 0) ||
			(!l._ascending && l.Compare(current1.value, current2.value) >= 0) {
			merged.Add(current1.value)
			current1 = current1.next
		} else {
			merged.Add(current2.value)
			current2 = current2.next
		}
	}

	for current1 != nil {
		merged.Add(current1.value)
		current1 = current1.next
	}

	for current2 != nil {
		merged.Add(current2.value)
		current2 = current2.next
	}

	l.head = merged.head
	l.tail = merged.tail

	return nil
}

func (l *OrderedList[T]) Contains(sublist []T) bool {
	if len(sublist) == 0 {
		return true
	}

	if l.head == nil {
		return false
	}

	current := l.head
	for current != nil {
		if l.Compare(current.value, sublist[0]) == 0 {
			if l.checkSublistFromNode(current, sublist) {
				return true
			}
		}
		current = current.next
	}
	return false
}

func (l *OrderedList[T]) checkSublistFromNode(startNode *Node[T], sublist []T) bool {
	current := startNode
	for i := 0; i < len(sublist); i++ {
		if current == nil || l.Compare(current.value, sublist[i]) != 0 {
			return false
		}
		current = current.next
	}
	return true
}

func (l *OrderedList[T]) FindMostFrequent() (T, int, error) {
	var zeroValue T
	if l.head == nil {
		return zeroValue, 0, errors.New("список пуст")
	}

	maxCount := 0
	maxValue := zeroValue
	currentCount := 1
	currentValue := l.head.value

	current := l.head.next
	for current != nil {
		if l.Compare(current.value, currentValue) == 0 {
			currentCount++
		} else {
			if currentCount > maxCount {
				maxCount = currentCount
				maxValue = currentValue
			}
			currentValue = current.value
			currentCount = 1
		}
		current = current.next
	}

	if currentCount > maxCount {
		maxCount = currentCount
		maxValue = currentValue
	}

	return maxValue, maxCount, nil
}

type IndexedOrderedList[T constraints.Ordered] struct {
	OrderedList[T]
	nodes []*Node[T]
}

func (l *IndexedOrderedList[T]) Add(item T) {
	l.OrderedList.Add(item)
	l.rebuildIndex()
}

func (l *IndexedOrderedList[T]) Delete(item T) {
	l.OrderedList.Delete(item)
	l.rebuildIndex()
}

func (l *IndexedOrderedList[T]) Clear(asc bool) {
	l.OrderedList.Clear(asc)
	l.nodes = nil
}

func (l *IndexedOrderedList[T]) rebuildIndex() {
	count := l.Count()
	l.nodes = make([]*Node[T], 0, count)
	current := l.head
	for current != nil {
		l.nodes = append(l.nodes, current)
		current = current.next
	}
}

func (l *IndexedOrderedList[T]) FindIndex(item T) (int, error) {
	count := l.Count()
	if count == 0 {
		return -1, errors.New("список пуст")
	}

	left, right := 0, count-1

	for left <= right {
		mid := (left + right) / 2
		cmp := l.Compare(l.nodes[mid].value, item)

		if cmp == 0 {
			for mid > 0 && l.Compare(l.nodes[mid-1].value, item) == 0 {
				mid--
			}
			return mid, nil
		} else if (l._ascending && cmp < 0) || (!l._ascending && cmp > 0) {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	return -1, errors.New("элемент не найден")
}

func (l *IndexedOrderedList[T]) GetByIndex(index int) (T, error) {
	var zeroValue T
	count := l.Count()
	if index < 0 || index >= count {
		return zeroValue, errors.New("индекс вне диапазона")
	}
	return l.nodes[index].value, nil
}

func (l *OrderedList[T]) ToSlice() []T {
	count := l.Count()
	result := make([]T, 0, count)
	current := l.head
	for current != nil {
		result = append(result, current.value)
		current = current.next
	}
	return result
}

func (l *OrderedList[T]) Print() {
	fmt.Print("Список: ")
	current := l.head
	for current != nil {
		fmt.Printf("%v ", current.value)
		current = current.next
	}
	fmt.Printf("(порядок: %t, количество: %d)\n", l._ascending, l.Count())
}

func NewOrderedList[T constraints.Ordered](ascending bool) *OrderedList[T] {
	list := &OrderedList[T]{}
	list.Clear(ascending)
	return list
}

func NewIndexedOrderedList[T constraints.Ordered](ascending bool) *IndexedOrderedList[T] {
	list := &IndexedOrderedList[T]{}
	list.Clear(ascending)
	return list
}

File: alg_ds_1/python_version/07_ordered_list/ordered_list_2.py
from ordered_list import Node, OrderedList as BaseOrderedList


class OrderedList(BaseOrderedList):

    def __init__(self, asc):
        super().__init__(asc)
        self._index = []

    def add(self, value):
        super().add(value)
        self._rebuild_index()

    def delete(self, val):
        super().delete(val)
        self._rebuild_index()

    def clean(self, asc):
        super().clean(asc)
        self._index = []

    # Задание на курсе: 7
    # Задача: 8
    # Название: удаление всех дубликатов
    # Временная сложность: O(n)
    # Пространственная сложность: O(n) для вспомогательного индекса
    def delete_duplicates(self):
        current = self.head

        while current is not None and current.next is not None:
            if self.compare(current.value, current.next.value) != 0:
                current = current.next
                continue

            self._unlink_node(current.next)

        self._rebuild_index()

    # Задание на курсе: 7
    # Задача: 9
    # Название: слияние двух упорядоченных списков
    # Временная сложность: O(n + m)
    # Пространственная сложность: O(n + m)
    def merge(self, other):
        if self._is_ascending() != other._is_ascending():
            raise ValueError('Lists must have the same sort order')

        result = OrderedList(self._is_ascending())
        left = self.head
        right = other.head

        while left is not None and right is not None:
            if self._should_take_left(left.value, right.value):
                result._append_indexed_value(left.value)
                left = left.next
                continue
            result._append_indexed_value(right.value)
            right = right.next

        while left is not None:
            result._append_indexed_value(left.value)
            left = left.next

        while right is not None:
            result._append_indexed_value(right.value)
            right = right.next

        return result

    # Задание на курсе: 7
    # Задача: 10
    # Название: проверка наличия упорядоченного подсписка
    # Временная сложность: O(n * m)
    # Пространственная сложность: O(1)
    def contains_sublist(self, sublist):
        if sublist.head is None:
            return True

        if self._is_ascending() != sublist._is_ascending():
            return False

        current = self.head

        while current is not None:
            comparison = self.compare(current.value, sublist.head.value)

            if comparison == 0 and self._matches_sublist(current, sublist.head):
                return True

            passed_value = (
                self._is_ascending() and comparison > 0
                or not self._is_ascending() and comparison < 0
            )

            if passed_value:
                return False

            current = current.next

        return False

    # Задание на курсе: 7
    # Задача: 11
    # Название: поиск наиболее часто встречающегося значения
    # Временная сложность: O(n)
    # Пространственная сложность: O(1)
    def find_most_frequent(self):
        if self.head is None:
            return None

        most_frequent = self.head.value
        maximum_count = 1
        current_value = self.head.value
        current_count = 1
        current = self.head.next

        while current is not None:
            if self.compare(current.value, current_value) == 0:
                current_count += 1
                current = current.next
                continue

            if current_count > maximum_count:
                most_frequent = current_value
                maximum_count = current_count

            current_value = current.value
            current_count = 1
            current = current.next

        if current_count > maximum_count:
            most_frequent = current_value

        return most_frequent

    # Задание на курсе: 7
    # Задача: 12
    # Название: поиск индекса элемента
    # Временная сложность: O(log n)
    # Пространственная сложность: O(1) для вызова, O(n) для индекса
    def find_index(self, value):
        left = 0
        right = len(self._index)

        while left < right:
            middle = (left + right) // 2
            comparison = self.compare(self._index[middle].value, value)
            middle_before_value = (
                self._is_ascending() and comparison < 0
                or not self._is_ascending() and comparison > 0
            )

            if middle_before_value:
                left = middle + 1
                continue

            right = middle

        if left == len(self._index):
            return -1

        if self.compare(self._index[left].value, value) != 0:
            return -1

        return left

    def _matches_sublist(self, current, expected):
        while expected is not None:
            if current is None:
                return False

            if self.compare(current.value, expected.value) != 0:
                return False

            current = current.next
            expected = expected.next

        return True

    def _append_indexed_value(self, value):
        node = Node(value)
        self._append_node(node)
        self._index.append(node)

    def _should_take_left(self, left_value, right_value):
        comparison = self.compare(left_value, right_value)

        if self._is_ascending():
            return comparison <= 0

        return comparison >= 0

    def _rebuild_index(self):
        self._index = self.get_all()

# Рефлексия по заданию 5
# Вращение сделал ровно так же, с вталкиванием/выталкиванием.
# С перемещением во второй стек только если он пустой неочевидна (лично для меня),
# но я поэтому и запомнил :)
# И в целом прием разворачивания не-массивов через стек кажется удобным.
# С круговой очередью описание (и задача) кажутся намного сложнее, чем по факту есть.
# А код написал - и норм.
